src/what-unsafe-does.md

```diff
@@ -16,19 +16,32 @@ to your program. You definitely *should not* invoke Undefined Behavior.
 Unlike C, Undefined Behavior is pretty limited in scope in Rust. All the core
 language cares about is preventing the following things:
 
-* Dereferencing null, dangling, or unaligned pointers
+* Loading from or storing to null, dangling, or unaligned references or raw
+  pointers
```

Comment:
I changed "dereferencing" into "loading from or storing to", because if a dangling/unaligned pointer is dereferenced (as in `*ptr` is syntactically written) without incurring a memory access (as in `&(*ptr).field`), that does not actually cause UB to my knowledge.

Reply:
That's a pretty subtle claim! I'm not sure the nomicon can justifiably document something so subtle. As a general rule, we just say more things are UB than *actually* are, because it is of course completely fine for devs to be more careful than the really need to be, and it's back-compat to suddenly say "this is now *NOT* UB"

Reply:
> As a general rule, we just say more things are UB than actually are

The problem is that this can lead to people just ignoring what we say because "it says too much UB anyway".

This is also somewhat related to https://github.com/rust-lang/rfcs/pull/2582.

Reply:
Except that you've also said yourself before, I think it was on Zulip recently: Until something says you _can_ do something, it's UB by default.

Reply:
That's *my* position, but my experience tell me other people take a more... "relaxed" approach to UB.

My main reason that I don't thing dereferencing a dangling pointer should be UB is that I hope to, some day, clean up [this mess](https://github.com/rust-lang/rust/blob/95b1fe560d2bd8472f250fb8cfd2168520a58405/src/liballoc/collections/btree/node.rs#L594-L609). But to do that I need to be able to take a raw pointer to a large type that is partially dangling, and still compute the address of a field in there that is in the non-dangling part.

Reply:
I think I came to the conclusion that for now the Nomicon should probably stick to "dereference", and I think that also means we don't need to "pointer offset" clause.

Some time after we resolved https://github.com/rust-lang/rfcs/pull/2582 (heh), we should officially determine if its "deref" or "access" that counts, and then adjust whatever else needs adjusting.